internal/olm/operator/internal/registry_pod.go

```diff
@@ -0,0 +1,276 @@
+		}
+		return p.Status.Phase == corev1.PodRunning, nil
+	})
+
+	err := wait.PollImmediateUntil(time.Duration(1*time.Second), podCheck, ctx.Done())
+	if err != nil {
+		return fmt.Errorf("error waiting for registry pod %s to run: %v", pod.Name, err)
+	}
+
+	_, err = rp.fetchPodLogs(ctx, pod)
+	if err != nil {
+		return fmt.Errorf("error in fetching pod logs: %v", err)
+	}
+
+	return err
+}
+
+func (rp *RegistryPod) init() error {
+	rp.setDefaults()
+	err := rp.validate()
+	if err != nil {
+		return fmt.Errorf("error in validating registry pod: %v", err)
+	}
+	return nil
+}
+
+func (rp *RegistryPod) setDefaults() {
+	// set the grpcPort to default 50051
+	rp.grpcPort = defaultGRPCPort
+
+	if len(rp.IndexImage) == 0 {
+		rp.IndexImage = defaultIndexImage
+	}
+
+	if len(rp.BundleAddMode) == 0 {
+		if rp.IndexImage == defaultIndexImage {
+			rp.BundleAddMode = "semver"
+		} else {
+			rp.BundleAddMode = "replaces"
+		}
+	}
+}
+
+func (rp *RegistryPod) validate() error {
+	if len(rp.BundleImage) == 0 {
+		return errors.New("bundle image is required")
+	}
+	if len(rp.DBPath) == 0 {
+		return errors.New("registry database path is required")
+	}
+
+	if len(rp.Namespace) == 0 {
+		return errors.New("pod namespace is required")
+	}
+
+	return nil
+}
+
+// getPodName will return a string constructed from the bundle Image name
+func getPodName(bundleName string) (string, error) {
+	var podName string
+	if strings.Contains(bundleName, "/") {
+		split := strings.Split(bundleName, "/")
+		lastSegment := strings.Split(split[len(split)-1:][0], ":")
+		// get the last but one element from last segment excluding
+		// the version number and append index
+		podName = lastSegment[len(lastSegment)-2] + "-index"
+	} else {
+		return "", fmt.Errorf("invalid bundle image name: %s", bundleName)
+	}
+	return podName, nil
+}
+
+// build sets the defaults and validates the RegistryPod struct, and returns a pod definition
+func (rp *RegistryPod) build() (*corev1.Pod, error) {
+	podName, err := getPodName(rp.BundleImage)
+	if err != nil {
+		return nil, fmt.Errorf("error in building pod: %v", err)
+	}
+
+	containerCmd, err := rp.getContainerCmd()
+	if err != nil {
+		return nil, fmt.Errorf("error in parsing container command: %v", err)
+	}
+
+	pod := &corev1.Pod{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      podName,
+			Namespace: rp.Namespace,
+		},
+		Spec: corev1.PodSpec{
+			Containers: []corev1.Container{
+				{
+					Name:  defaultContainerName,
+					Image: rp.IndexImage,
+					Command: []string{
+						"/bin/sh",
+						"-c",
+						containerCmd,
+					},
+					Ports: []corev1.ContainerPort{
+						{Name: defaultContainerPortName, ContainerPort: rp.grpcPort},
+					},
+				},
+			},
+		},
+	}
+
+	return pod, nil
+}
+
+func (rp *RegistryPod) getContainerCmd() (string, error) {
+	const containerCommand = `
+/bin/mkdir -p /database && 
+/bin/opm registry add -d {{.DBPath}} -b {{.BundleImage}} --mode={{.BundleAddMode}} &&
+/bin/opm registry serve -d {{.DBPath}}
```

Comment:
We need to tell `serve` what port to listen on, so it aligns with how we setup the container ports.
```suggestion
/bin/opm registry serve -d {{.DBPath}} -p {{.GRPCPort}}
```